tracer/src/Datadog.Trace/Logging/DirectSubmission/Formatting/LogFormatter.cs

```diff
@@ -237,5 +239,91 @@ internal static JsonTextWriter GetJsonWriter(StringBuilder builder)
 
             writer.WriteEndObject();
         }
+
+        internal void FormatCIVisibilityLog(StringBuilder sb, string source, string? logLevel, string message, ISpan? span)
+        {
+            using var writer = GetJsonWriter(sb);
+
+            // Based on JsonFormatter
+            writer.WriteStartObject();
+
+            writer.WritePropertyName("ddsource", escape: false);
+            writer.WriteValue(source);
+
+            if (_host is not null)
+            {
+                writer.WritePropertyName("hostname", escape: false);
+                writer.WriteValue(_host);
+            }
+
+            writer.WritePropertyName("timestamp", escape: false);
+            writer.WriteValue(DateTimeOffset.UtcNow.ToUnixTimeNanoseconds() / 1_000_000);
+
+            if (logLevel is not null)
+            {
+                writer.WritePropertyName("status", escape: false);
+                writer.WriteValue(logLevel);
+            }
+
+            writer.WritePropertyName("message", escape: false);
+            writer.WriteValue(message);
+
+            var env = _env ?? string.Empty;
+            var service = _service;
+            if (span is not null)
+            {
+                if (span.GetTag(Tags.Env) is { } spanEnv)
+                {
+                    env = spanEnv;
+                }
+
+                if (!string.IsNullOrEmpty(span.ServiceName))
+                {
+                    service = span.ServiceName;
+                }
+
+                writer.WritePropertyName("dd.trace_id", escape: false);
+                writer.WriteValue($"{span.TraceId}");
+
+                writer.WritePropertyName("dd.span_id", escape: false);
+                writer.WriteValue($"{span.SpanId}");
+
+                if (span.GetTag(TestTags.Suite) is { } suite)
+                {
+                    writer.WritePropertyName(TestTags.Suite, escape: false);
+                    writer.WriteValue(suite);
+                }
+
+                if (span.GetTag(TestTags.Name) is { } name)
+                {
+                    writer.WritePropertyName(TestTags.Name, escape: false);
+                    writer.WriteValue(name);
+                }
+
+                if (span.GetTag(TestTags.Bundle) is { } bundle)
+                {
+                    writer.WritePropertyName(TestTags.Bundle, escape: false);
+                    writer.WriteValue(bundle);
+                }
+            }
+
+            writer.WritePropertyName("service", escape: false);
+            writer.WriteValue(service);
+
+            // spaces are not allowed inside ddtags
+            env = env.Replace(" ", string.Empty);
+            env = env.Replace(":", string.Empty);
+
+            var ddtags = $"env:{env},datadog.product:citest";
+            if (_globalTags is { Length: > 0 } globalTags)
+            {
+                ddtags += "," + globalTags;
+            }
```

Comment:
Minor optimisation - if `span.GetTag(Tags.Env) is null` then you could pre-cache all this concatenating in the constructor, as the rest of it is static?

Reply:
Ok, I did the optimisation in https://github.com/DataDog/dd-trace-dotnet/pull/3037/commits/1511ea663385cb7e0f4ca38d5add5791dcadba16, but avoiding the `.ctor` I don't like the idea of CIVisibility adding an extra allocation on APM.